immuneML/encodings/pdbencoding/PDBEncoder.py

```diff
@@ -0,0 +1,250 @@
+from immuneML.data_model.encoded_data.EncodedData import EncodedData
+
+from immuneML.data_model.dataset.PDBDataset import PDBDataset
+from immuneML.encodings.DatasetEncoder import DatasetEncoder
+from immuneML.encodings.EncoderParams import EncoderParams
+import numpy as np
+from immuneML.data_model.receptor.RegionType import RegionType
+
+
+class PDBEncoder(DatasetEncoder):
+
+
+
+    @staticmethod
+    def build_object(dataset=None, **params):
+        return PDBEncoder(**params)
+
+    def __init__(self, name: str = None, region_type: RegionType = None):
+
+        self.name = name
+        self.region_type = region_type
+
+    def encode(self, dataset, params: EncoderParams):
+
+        structures =[]
+
+        pdbParser = PDBParser(
+            PERMISSIVE=True
+        )
+
+        CDRcounter = 0
+
+
+
+        for files in dataset.get_data():
+            parserObject = pdbParser.get_structure("pdbStructure", files)
+
+
+            collection = []
+            list = []
+
+            region = self.region_type.name
+            counter = 0
+
+            for model in parserObject:
+                for chain in model:
+                    residueCounter = 0
+                    for residue in chain:
+                        if residueCounter == 0:
+                            list.append(residue.get_parent())
+                            residueCounter = 1
+
+                        for atom in residue:
+                            if (atom.get_name() == "CA"):
+
+                                if "FULL_SEQUENCE" in region:
+                                    list.append(atom.get_coord())
+
+                                else:
+                                    if "IMGT_CDR3" in region and (isInCDR3(atom.full_id[3][1]) or len(collection) >= 2):
+                                        list.append(atom.get_coord())
+
+                                    if "IMGT_CDR2" in region and (isInCDR2(atom.full_id[3][1]) or len(collection) >= 2):
+                                        list.append(atom.get_coord())
+
+                                    if "IMGT_CDR1" in region and (isInCDR1(atom.full_id[3][1]) or len(collection) >= 2):
+                                        list.append(atom.get_coord())
+
+
+                    saveList = list.copy()
+                    collection.append(saveList)
+                    counter = counter + 1
+                    list.clear()
+
+            lightID = -1
+            heavyID = -1
+            antigenID = -1
+
+            for i in range(0,len(collection)):
+                chainID = collection[i][0].id
+
+                if 'L' in chainID or 'A' in chainID:
+                    lightID = i
+
+                elif 'H' in chainID or 'B' in chainID:
+                    heavyID = i
+
+                elif 'P' in chainID or 'C' in chainID:
+                    antigenID = i
+
+                collection[i].pop(0)
+
+
+            if lightID >= 0 and heavyID >= 0 and antigenID == -1:
```

Comment:
Some of your methods are very large, which makes them harder to read and more prone to bugs. Smaller methods, especially when they have nice descriptive names, are always better (a function of a single line is absolutely not too small!). Clean code book chapter 3 is an amazing resource for this https://thixalongmy.haugiang.gov.vn/media/1175/clean_code.pdf 

There are already some low-hanging fruits here: some parts of this large methods you already separated a bit by having some spacing before and after. They can easily be put in a method by themselves with a nice descriptive name and the flow of the code would be much easier to follow.

Reply:
in other places it may be less trivial to break all the nested if statements and for loops apart. but those parts of the code can usually benefit the most from being refactored, and maybe the algorithm needs to change a bit to be split up into smaller chunks.